OpenRA.Mods.Common/Widgets/SupportPowerTimerWidget.cs

```diff
@@ -11,55 +11,148 @@
+		readonly string fontName = "Bold";
+		readonly SpriteFont font;
+		readonly float yIncrement, powerPos, timePos;
+
+		// Represents one line of text
+		struct Text
 		{
-			powers = world.ActorsWithTrait<SupportPowerManager>()
-				.Where(p => !p.Actor.IsDead && !p.Actor.Owner.NonCombatant)
-				.SelectMany(s => s.Trait.Powers.Values)
-				.Where(p => p.Instances.Any() && p.Info.DisplayTimer && !p.Disabled);
+			public Color Color;
+			public string Team;		// Used only for team label
+			public string Player;
+			public string Power;
+			public string RemainingTime;
 		}
 
-		public override void Tick()
+		readonly List<Text> display;
+		readonly bool noPowers;
+		bool alternateColor, needTeamLabel, showPlayer;
+
+		readonly World world;
+		int lastTick;
+
+		[ObjectCreator.UseCtor]
+		public SupportPowerTimerWidget(World world)
 		{
-			texts = powers.Select(p =>
-			{
-				var time = WidgetUtils.FormatTime(p.RemainingTime, false);
-				var text = Format.F(p.Info.Description, time);
-				var color = !p.Ready || Game.LocalTick % 50 < 25 ? p.Instances[0].Self.Owner.Color.RGB : Color.White;
-				return Pair.New(text, color);
-			}).ToArray();
+			spManagers = new Dictionary<Player, SupportPowerManager>();
+			foreach (var tp in world.ActorsWithTrait<SupportPowerManager>()
+				.Where(tp => !tp.Actor.IsDead && !tp.Actor.Owner.NonCombatant))
+				spManagers[tp.Actor.Owner] = tp.Trait;
+
+			var availablePowers = world.Map.Rules.Actors.Values
+				.SelectMany(ai => ai.Traits.WithInterface<SupportPowerInfo>())
+				.Where(i => i.DisplayTimer).ToArray();
+
+			noPowers = !spManagers.Keys.Any() || !availablePowers.Any();
+			if (noPowers)
+				return;
+
+			playersByTeam = new SortedDictionary<int, Player[]>();
+			foreach (var team in spManagers.Keys.GroupBy(p => world.LobbyInfo.ClientWithIndex(p.ClientIndex).Team))
+				playersByTeam[team.Key] = team.ToArray();
+
+			font = Game.Renderer.Fonts[fontName];
+			yIncrement = font.Measure(" ").Y + 5;
+			var padding = font.Measure("  ").X;
+			var maxPlayerWidth = spManagers.Keys.Max(p => font.Measure(p.PlayerName).X);
+			var maxPowerWidth = availablePowers.Max(i => font.Measure(i.Description).X);
+			powerPos = maxPlayerWidth + padding;
+			timePos = powerPos + maxPowerWidth + padding;
+
+			display = new List<Text>();
+			this.world = world;
 		}
 
 		public override void Draw()
 		{
-			if (!IsVisible() || texts == null)
+			if (!IsVisible() || noPowers)
 				return;
 
-			var y = 0;
-			foreach (var t in texts)
+			// Generate texts once in every 10 WorldTicks,
+			// which is 0.4 seconds in game time.
+			if (world.WorldTick - lastTick > 9)
 			{
-				var font = Game.Renderer.Fonts[Font];
-				font.DrawTextWithContrast(t.First, new float2(Bounds.Location) + new float2(0, y), t.Second, Color.Black, 1);
-				y += (font.Measure(t.First).Y + 5) * (int)Order;
+				lastTick = world.WorldTick;
+				display.Clear();
+				alternateColor = Game.LocalTick % 50 > 25;
```

Comment:
I think 50 and 25 were used because local tick frequency is 25 per second, if there is no lag. So this would ideally show player color and flashing color for one second each.